tensorflow_datasets/structured/tic_tac_toe.py

```diff
@@ -0,0 +1,73 @@
+"""Binary classification task on possible configurations of tic-tac-toe game"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import tensorflow_datasets.public_api as tfds
+
+_CITATION = """
+@misc{Dua:2019 ,
+author = "Dua, Dheeru and Graff, Casey",
+year = "2017",
+title = "{UCI} Machine Learning Repository",
+url = "http://archive.ics.uci.edu/ml",
+institution = "University of California, Irvine, School of Information and Computer Sciences" }
+"""
+
+_DESCRIPTION = """
+This database encodes the complete set of possible board configurations at the end of tic-tac-toe games, where "x" is assumed to have played first. The target concept is "win for x" (i.e., true when "x" has one of 8 possible ways to create a "three-in-a-row").
+Interestingly, this raw database gives a stripped-down decision tree algorithm (e.g., ID3) fits. However, the rule-based CN2 algorithm, the simple IB1 instance-based learning algorithm, and the CITRE feature-constructing decision tree algorithm perform well on it.
+"""
+
+def convert(d):
+  if d == "o":
+    return 0
+  if d == "x":
+    return 1
+  if d == "b":
+    return 2
+
+import csv
+import tensorflow as tf
+import tensorflow_datasets.public_api as tfds
+
+class TicTacToe(tfds.core.GeneratorBasedBuilder):
+  """Binary classification task on possible configurations of tic-tac-toe game"""
+
+  VERSION = tfds.core.Version('2.0.0')
+
+  def _info(self):
+    return tfds.core.DatasetInfo(
+        builder=self,
+        description=_DESCRIPTION,
+        features=tfds.features.FeaturesDict({
+            "label": tfds.features.ClassLabel(names=["positive", "negative"]),
+            "features": tfds.features.Tensor(shape=(9,), dtype=tf.string)
```

Comment:
Each feature should be a `ClassLabel(names=["x", "o", "b"])` I think.

Reply:
No, as they are strings in the dataset, and I have converted them into integers using the `convert() ` function, so they have to be a Tensor which I can assign a `dtype ` to.

Reply:
Would it not be better to preserve the data as it is in the dataset, rather than converting it to integers?

Reply:
Not always, as sometimes adding this functionality makes the dataset much more intuitive. You can look at [titanic.py](https://github.com/tensorflow/datasets/blob/master/tensorflow_datasets/structured/titanic.py) for an example of this.

Reply:
Yes, observe that the classes are all labeled or converted to labels. Only the attributes that are originally quantitative values are of type `int` or `float`.

Reply:
Any use case would likely involve conversion to integers, hence I added this feature.

Reply:
Hm, I understand your point of view, however I do think it's better to leave the labels the way they are for better understanding.

Reply:
Alright,but I would like to wait for a mentor/reviewer's opinion. I do think what you're saying makes sense as well in that we stick to presenting the data as is, but I want to know if one extra function could potentially make importing and using this dataset easier.

Reply:
Sounds good :+1:

Reply:
This feature correspond to the disposition of labels (x, o, b) on a 3x3 grid.

Ideally, the feature correspond the most would be the LabeledImage proposal from this feature request (https://github.com/tensorflow/datasets/issues/910):
`LabeledImage(shape=(3, 3, 1), names=["b", "x", "o"])`

Unfortunately this is not available yet. The next best choices would be to choose between the `Label` or `Image` feature:
1. `Sequence(ClassLabel(names=["b", "x", "o"]), length=3*3)`
2. `Image(shape=(3, 3, 1))`

I'm in favour of 1., as it is can be converted trivially into image by the user with a `tf.reshape(labels, (3, 3, 1))`. With 2, the labels are lost for the users, which makes the example more difficult to interpret the image.

The user should be able to understand what each number represent. So I think it is important to keep the labels. The conversion str -> int is done automatically by TFDS, but the user can still recover the conversion int -> str with `info.features['labels'].int2str('x')`

The order of the labels is also important. The first label will have the value `0`, which in Python is associated with `False`:
1. tfds.features.ClassLabel(names=["positive", "negative"])  # positive=0, negative=1
2. tfds.features.ClassLabel(names=["negative", "positive"])  # positive=1, negative=2

2. is a much better choice because it match the user expectation that: `bool(str2int('negative')) == False` and `bool(str2int('positive')) == True`, which is not the case here.

Similarly, for the grid, `x` and `o` correspond to play, while `b` means "background"/"default"/"empty". In this case it make sense for the first label to be `b`, as an empty grid would just be a grid of 0.

Reply:
Thanks for the insight @Conchylicultor

Reply:
Thank you @Conchylicultor, I'll make sure to implement the changes ASAP.